### DISCUSSION TITLE
Overview of Actions and Transactions

### DISCUSSION THREAD

**Discussion by martineckardt** [ava-labs/hypersdk#16] [Show and Tell]
I prepared a few images explaining the concepts of actions, transactions and upgradeability:

# Actions

From what I saw in the code and documentation I understand that actions define a set of different actions user can submit using transactions. Each Action has a ActionHandler / execute function that defines the business logic.

<img width="1308" alt="action" src="https://user-images.githubusercontent.com/10773395/220381228-f94dcdba-71ab-43ae-b414-8daafb72dd10.png">

# Transactions

What really stood out to me is the account abstraction, meaning that the sdk is very flexible how account are created. If compared to a transaction of the SubnetEVM, transactions are very light weight and flexible.

<img width="1293" alt="image" src="https://user-images.githubusercontent.com/10773395/220382967-a0cf2de1-a888-481e-acac-1c7bd3c1d85a.png">

Source: [ava-labs/subnet-evm](https://github.com/ava-labs/subnet-evm/blob/4209aa4a25b18d3e90560091a14660f664855a2f/core/types/transaction.go#L61), [ava-labs/hypersdk](https://github.com/ava-labs/hypersdk/blob/fd5e14c3a1905c5554105d62911249499e68a8ee/chain/transaction.go#L25)

# Upgradeability

Each action is scoped by block timestamps and therefore they can be replaced by newer versions or disabled. 

<img width="1300" alt="upgrades" src="https://user-images.githubusercontent.com/10773395/220381229-8badf6c3-b4f1-49e8-baad-71640306ee22.png">

**Comment by patrick-ogrady** [comment]
Thanks for sharing, @martineckardt!
